gladier_xpcs/flow_gpu.py

```diff
@@ -0,0 +1,63 @@
+import logging
+import time
+import random
+import globus_sdk
+from gladier import GladierBaseClient, generate_flow_definition, utils
+# import gladier_xpcs.log  # Uncomment for debug logging
+
+
+log = logging.getLogger(__name__)
+
+@generate_flow_definition(modifiers={
+   'publish_gather_metadata': {'payload': '$.GatherXpcsMetadata.details.result[0]'}
+})
+class XPCSGPUFlow(GladierBaseClient):
+    globus_group = '368beb47-c9c5-11e9-b455-0efb3ba9a670'
+    max_retries = 10
+    containers = {}
+
+    gladier_tools = [
+        'gladier_xpcs.tools.TransferFromClutchToTheta',
+        'gladier_xpcs.tools.PrePublish',
+        'gladier_xpcs.tools.AcquireNodes',
+        'gladier_xpcs.tools.EigenCorrGPU',
+        'gladier_xpcs.tools.MakeCorrPlots',
+        'gladier_xpcs.tools.gather_xpcs_metadata.GatherXPCSMetadata',
+        'gladier_xpcs.tools.Publish',
+    ]
+
+    def register_funcx_function(self, function):
+        """Register containers for any functions listed in self.containers"""
+        fxid_name = utils.name_generation.get_funcx_function_name(function)
+        if fxid_name in self.containers.keys():
+            container = self.containers[fxid_name]
+            log.info(f'Registering {fxid_name} with {container["location"]}')
+            fxid_name = utils.name_generation.get_funcx_function_name(function)
+            fxck_name = utils.name_generation.get_funcx_function_checksum_name(function)
+            cfg = self.get_cfg(private=True)
+            cid = self.funcx_client.register_container(**container)
+            fxid = self.funcx_client.register_function(function, function.__doc__, container_uuid=cid)
+            cfg[self.section][fxid_name] = fxid
+            cfg[self.section][fxck_name] = self.get_funcx_function_checksum(function)
+            cfg.save()
+        else:
+            super().register_funcx_function(function)
+
+    def retry_backoff(self, method, *args, **kwargs):
+        retries = 0
+        while retries < self.max_retries:
+            try:
+                return method(*args, **kwargs)
+            except globus_sdk.GlobusTimeoutError:
+                time.sleep(random.randint(1, 10))
+            except globus_sdk.GlobusAPIError as gapie:
+                if gapie.http_status == 400:
+                    raise
+                time.sleep(random.randint(1, 10))
+        raise
+
+    def run_flow(self, *args, **kwargs):
+        return self.retry_backoff(super().run_flow, *args, **kwargs)
+
+    def get_status(self, *args, **kwargs):
+        return self.retry_backoff(super().get_status, *args, **kwargs)
```

Comment:
We might consider moving the duplicated code into a base class for re-usability on xpcs flows, especially if we add more clients to kick off reprocessing jobs. No worries on that now, but I think it would be a good future task.

Reply:
Agreed. Just saw this comment.